File: runtime/Dart/lib/src/atn/src/semantic_context.dart
/*
 * Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */

import 'package:antlr4/src/atn/src/atn_simulator.dart';
import 'package:collection/collection.dart';

import '../../recognizer.dart';
import '../../rule_context.dart';
import '../../util/murmur_hash.dart';

/// A tree structure used to record the semantic context in which
///  an ATN configuration is valid.  It's either a single predicate,
///  a conjunction {@code p1&&p2}, or a sum of products {@code p1||p2}.
///
///  <p>I have scoped the [AND], [OR], and [Predicate] subclasses of
///  [SemanticContext] within the scope of this outer class.</p>
abstract class SemanticContext {
  const SemanticContext();

  /// For context independent predicates, we evaluate them without a local
  /// context (i.e., null context). That way, we can evaluate them without
  /// having to create proper rule-specific context during prediction (as
  /// opposed to the parser, which creates them naturally). In a practical
  /// sense, this avoids a cast exception from RuleContext to myruleContext.
  ///
  /// <p>For context dependent predicates, we must pass in a local context so that
  /// references such as $arg evaluate properly as _localctx.arg. We only
  /// capture context dependent predicates in the context in which we begin
  /// prediction, so we passed in the outer context here in case of context
  /// dependent predicate evaluation.</p>
  bool eval(Recognizer parser, RuleContext? parserCallStack);

  /// Evaluate the precedence predicates for the context and reduce the result.
  ///
  /// @param parser The parser instance.
  /// @param parserCallStack
  /// @return The simplified semantic context after precedence predicates are
  /// evaluated, which will be one of the following values.
  /// <ul>
  /// <li>{@link #NONE}in if the predicate simplifies to [true] after
  /// precedence predicates are evaluated.</li>
  /// <li>nullin if the predicate simplifies to [false] after
  /// precedence predicates are evaluated.</li>
  /// <li>[this]in if the semantic context is not changed as a result of
  /// precedence predicate evaluation.</li>
  /// <li>A non-null [SemanticContext]in the new simplified
  /// semantic context after precedence predicates are evaluated.</li>
  /// </ul>
  SemanticContext? evalPrecedence(
    Recognizer parser,
    RuleContext? parserCallStack,
  ) {
    return this;
  }

  static SemanticContext? and(SemanticContext? a, SemanticContext? b) {
    if (a == null || a == EmptySemanticContext.Instance) return b;
    if (b == null || b == EmptySemanticContext.Instance) return a;
    final result = AND(a, b);
    if (result.opnds.length == 1) {
      return result.opnds[0];
    }

    return result;
  }

  ///
  ///  @see ParserATNSimulator#getPredsForAmbigAlts
  static SemanticContext? or(SemanticContext? a, SemanticContext? b) {
    if (a == null) return b;
    if (b == null) return a;
    if (a == EmptySemanticContext.Instance || b == EmptySemanticContext.Instance) return EmptySemanticContext.Instance;
    final result = OR(a, b);
    if (result.opnds.length == 1) {
      return result.opnds[0];
    }

    return result;
  }

  static Iterable<PrecedencePredicate> filterPrecedencePredicates(
      Iterable<SemanticContext> collection) {
    return collection.whereType<PrecedencePredicate>();
  }

  static Iterable<SemanticContext> filterNonPrecedencePredicates(
      Iterable<SemanticContext> collection) {
    return collection.where((e) => e is! PrecedencePredicate);
  }
}

class EmptySemanticContext extends SemanticContext {
  /// The default [SemanticContext], which is semantically equivalent to
  /// a predicate of the form {@code {true}?}.
  static const SemanticContext Instance = Predicate();

  @override
  bool eval(Recognizer<ATNSimulator> parser, RuleContext? parserCallStack) {
    return false;
  }
}

class Predicate extends SemanticContext {
  final int ruleIndex;
  final int predIndex;
  final bool isCtxDependent; // e.g., $i ref in pred

  const Predicate(
      [this.ruleIndex = -1, this.predIndex = -1, this.isCtxDependent = false]);

  @override
  bool eval(Recognizer parser, RuleContext? parserCallStack) {
    final localctx = isCtxDependent ? parserCallStack : null;
    return parser.sempred(localctx, ruleIndex, predIndex);
  }

  @override
  int get hashCode {
    var hashCode = MurmurHash.initialize();
    hashCode = MurmurHash.update(hashCode, ruleIndex);
    hashCode = MurmurHash.update(hashCode, predIndex);
    hashCode = MurmurHash.update(hashCode, isCtxDependent ? 1 : 0);
    hashCode = MurmurHash.finish(hashCode, 3);
    return hashCode;
  }

  @override
  bool operator ==(Object obj) {
    return obj is Predicate &&
        ruleIndex == obj.ruleIndex &&
        predIndex == obj.predIndex &&
        isCtxDependent == obj.isCtxDependent;
  }

  @override
  String toString() {
    return '{$ruleIndex:$predIndex}?';
  }
}

class PrecedencePredicate extends SemanticContext
    implements Comparable<PrecedencePredicate> {
  final int precedence;

  PrecedencePredicate([this.precedence = 0]);

  @override
  bool eval(Recognizer parser, RuleContext? parserCallStack) {
    return parser.precpred(parserCallStack, precedence);
  }

  @override
  SemanticContext? evalPrecedence(
    Recognizer parser,
    RuleContext? parserCallStack,
  ) {
    if (parser.precpred(parserCallStack, precedence)) {
      return EmptySemanticContext.Instance;
    } else {
      return null;
    }
  }

  @override
  int compareTo(PrecedencePredicate o) {
    return precedence - o.precedence;
  }

  @override
  int get hashCode {
    var hashCode = 1;
    hashCode = 31 * hashCode + precedence;
    return hashCode;
  }

  @override
  bool operator ==(Object other) {
    if (other is! PrecedencePredicate) {
      return false;
    }
    return precedence == other.precedence;
  }

// precedence >= _precedenceStack.peek()
  @override
  String toString() {
    return '{$precedence>=prec}?';
  }
}

/// This is the base class for semantic context "operators", which operate on
/// a collection of semantic context "operands".
///
/// @since 4.3
abstract class Operator extends SemanticContext {
  /// Gets the operands for the semantic context operator.
  ///
  /// @return a collection of [SemanticContext] operands for the
  /// operator.
  ///
  /// @since 4.3
  List<SemanticContext> get operands;
}

/// A semantic context which is true whenever none of the contained contexts
/// is false.

class AND extends Operator {
  late final List<SemanticContext> opnds;

  AND(SemanticContext a, SemanticContext b) {
    var operands = <SemanticContext>{};
    if (a is AND) {
      operands.addAll(a.opnds);
    } else {
      operands.add(a);
    }
    if (b is AND) {
      operands.addAll(b.opnds);
    } else {
      operands.add(b);
    }

    final precedencePredicates =
        SemanticContext.filterPrecedencePredicates(operands);

    operands = SemanticContext.filterNonPrecedencePredicates(operands).toSet();
    if (precedencePredicates.isNotEmpty) {
      // interested in the transition with the lowest precedence
      final reduced =
          precedencePredicates.reduce((a, b) => a.compareTo(b) <= 0 ? a : b);
      operands.add(reduced);
    }

    opnds = operands.toList();
  }

  @override
  List<SemanticContext> get operands {
    return opnds;
  }

  @override
  bool operator ==(Object other) {
    if (other is! AND) return false;
    return ListEquality().equals(opnds, other.opnds);
  }

  @override
  int get hashCode {
    return MurmurHash.getHashCode(opnds, runtimeType.hashCode);
  }

  /// {@inheritDoc}
  ///
  /// <p>
  /// The evaluation of predicates by this context is short-circuiting, but
  /// unordered.</p>

  @override
  bool eval(Recognizer parser, RuleContext? parserCallStack) {
    for (var opnd in opnds) {
      if (!opnd.eval(parser, parserCallStack)) return false;
    }
    return true;
  }

  @override
  SemanticContext? evalPrecedence(
    Recognizer parser,
    RuleContext? parserCallStack,
  ) {
    var differs = false;
    final operands = <SemanticContext>[];
    for (var context in opnds) {
      final evaluated = context.evalPrecedence(parser, parserCallStack);
      differs |= (evaluated != context);
      if (evaluated == null) {
        // The AND context is false if any element is false
        return null;
      } else if (evaluated != EmptySemanticContext.Instance) {
        // Reduce the result by skipping true elements
        operands.add(evaluated);
      }
    }

    if (!differs) {
      return this;
    }

    if (operands.isEmpty) {
      // all elements were true, so the AND context is true
      return EmptySemanticContext.Instance;
    }

    SemanticContext? result = operands[0];
    for (var i = 1; i < operands.length; i++) {
      result = SemanticContext.and(result, operands[i]);
    }

    return result;
  }

  @override
  String toString() {
    return opnds.join('&&');
  }
}

/// A semantic context which is true whenever at least one of the contained
/// contexts is true.
class OR extends Operator {
  late final List<SemanticContext> opnds;

  OR(SemanticContext a, SemanticContext b) {
    var operands = <SemanticContext>{};
    if (a is OR) {
      operands.addAll(a.opnds);
    } else {
      operands.add(a);
    }
    if (b is OR) {
      operands.addAll(b.opnds);
    } else {
      operands.add(b);
    }

    final precedencePredicates =
        SemanticContext.filterPrecedencePredicates(operands);

    operands = SemanticContext.filterNonPrecedencePredicates(operands).toSet();
    if (precedencePredicates.isNotEmpty) {
      // interested in the transition with the highest precedence
      final reduced =
          precedencePredicates.reduce((a, b) => a.compareTo(b) >= 0 ? a : b);
      operands.add(reduced);
    }

    opnds = operands.toList();
  }

  @override
  List<SemanticContext> get operands {
    return opnds;
  }

  @override
  bool operator ==(Object other) {
    if (other is! OR) return false;
    return ListEquality().equals(opnds, other.opnds);
  }

  @override
  int get hashCode {
    return MurmurHash.getHashCode(opnds, runtimeType.hashCode);
  }

  /// {@inheritDoc}
  ///
  /// <p>
  /// The evaluation of predicates by this context is short-circuiting, but
  /// unordered.</p>

  @override
  bool eval(Recognizer parser, RuleContext? parserCallStack) {
    for (var opnd in opnds) {
      if (opnd.eval(parser, parserCallStack)) return true;
    }
    return false;
  }

  @override
  SemanticContext? evalPrecedence(
    Recognizer parser,
    RuleContext? parserCallStack,
  ) {
    var differs = false;
    final operands = <SemanticContext>[];
    for (var context in opnds) {
      final evaluated = context.evalPrecedence(parser, parserCallStack);
      differs |= (evaluated != context);
      if (evaluated == EmptySemanticContext.Instance) {
        // The OR context is true if any element is true
        return EmptySemanticContext.Instance;
      } else if (evaluated != null) {
        // Reduce the result by skipping false elements
        operands.add(evaluated);
      }
    }

    if (!differs) {
      return this;
    }

    if (operands.isEmpty) {
      // all elements were false, so the OR context is false
      return null;
    }

    SemanticContext? result = operands[0];
    for (var i = 1; i < operands.length; i++) {
      result = SemanticContext.or(result, operands[i]);
    }

    return result;
  }

  @override
  String toString() {
    return opnds.join('||');
  }
}

File: runtime/CSharp/src/Atn/SemanticContext.cs
/* Copyright (c) 2012-2017 The ANTLR Project. All rights reserved.
 * Use of this file is governed by the BSD 3-clause license that
 * can be found in the LICENSE.txt file in the project root.
 */
using System;
using System.Collections.Generic;
using System.Linq;
using Antlr4.Runtime.Misc;
using Antlr4.Runtime.Sharpen;

namespace Antlr4.Runtime.Atn
{
    public abstract class SemanticContext
    {
        public abstract bool Eval<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            where ATNInterpreter : ATNSimulator;

		public virtual SemanticContext EvalPrecedence<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            where ATNInterpreter : ATNSimulator
        {
            return this;
        }

        public class Empty : SemanticContext
        {
            public static readonly SemanticContext Instance = new Empty();

            public override bool Eval<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            {
                return false;
            }
        }

        public class Predicate : SemanticContext
        {
            public readonly int ruleIndex;

            public readonly int predIndex;

            public readonly bool isCtxDependent;

            protected internal Predicate()
            {
                // e.g., $i ref in pred
                this.ruleIndex = -1;
                this.predIndex = -1;
                this.isCtxDependent = false;
            }

            public Predicate(int ruleIndex, int predIndex, bool isCtxDependent)
            {
                this.ruleIndex = ruleIndex;
                this.predIndex = predIndex;
                this.isCtxDependent = isCtxDependent;
            }

            public override bool Eval<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            {
                RuleContext localctx = isCtxDependent ? parserCallStack : null;
                return parser.Sempred(localctx, ruleIndex, predIndex);
            }

            public override int GetHashCode()
            {
                int hashCode = MurmurHash.Initialize();
                hashCode = MurmurHash.Update(hashCode, ruleIndex);
                hashCode = MurmurHash.Update(hashCode, predIndex);
                hashCode = MurmurHash.Update(hashCode, isCtxDependent ? 1 : 0);
                hashCode = MurmurHash.Finish(hashCode, 3);
                return hashCode;
            }

            public override bool Equals(object obj)
            {
                if (!(obj is SemanticContext.Predicate))
                {
                    return false;
                }
                if (this == obj)
                {
                    return true;
                }
                SemanticContext.Predicate p = (SemanticContext.Predicate)obj;
                return this.ruleIndex == p.ruleIndex && this.predIndex == p.predIndex && this.isCtxDependent == p.isCtxDependent;
            }

            public override string ToString()
            {
                return "{" + ruleIndex + ":" + predIndex + "}?";
            }
        }

        public class PrecedencePredicate : SemanticContext, IComparable<SemanticContext.PrecedencePredicate>
        {
            public readonly int precedence;

            protected internal PrecedencePredicate()
            {
                this.precedence = 0;
            }

            public PrecedencePredicate(int precedence)
            {
                this.precedence = precedence;
            }

            public override bool Eval<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            {
                return parser.Precpred(parserCallStack, precedence);
            }

            public override SemanticContext EvalPrecedence<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            {
                if (parser.Precpred(parserCallStack, precedence))
                {
                    return SemanticContext.Empty.Instance;
                }
                else
                {
                    return null;
                }
            }

            public virtual int CompareTo(SemanticContext.PrecedencePredicate o)
            {
                return precedence - o.precedence;
            }

            public override int GetHashCode()
            {
                int hashCode = 1;
                hashCode = 31 * hashCode + precedence;
                return hashCode;
            }

            public override bool Equals(object obj)
            {
                if (!(obj is SemanticContext.PrecedencePredicate))
                {
                    return false;
                }
                if (this == obj)
                {
                    return true;
                }
                SemanticContext.PrecedencePredicate other = (SemanticContext.PrecedencePredicate)obj;
                return this.precedence == other.precedence;
            }

            public override string ToString()
            {
                // precedence >= _precedenceStack.peek()
                return "{" + precedence + ">=prec}?";
            }
        }

        public abstract class Operator : SemanticContext
        {
            [NotNull]
            public abstract ICollection<SemanticContext> Operands
            {
                get;
            }
        }

        public class AND : SemanticContext.Operator
        {
            [NotNull]
            public readonly SemanticContext[] opnds;

            public AND(SemanticContext a, SemanticContext b)
            {
                HashSet<SemanticContext> operands = new HashSet<SemanticContext>();
                if (a is SemanticContext.AND)
                {
                    operands.UnionWith(((AND)a).opnds);
                }
                else
                {
                    operands.Add(a);
                }
                if (b is SemanticContext.AND)
                {
                    operands.UnionWith(((AND)b).opnds);
                }
                else
                {
                    operands.Add(b);
                }
                IList<SemanticContext.PrecedencePredicate> precedencePredicates = FilterPrecedencePredicates(operands);
                if (precedencePredicates.Count > 0)
                {
                    // interested in the transition with the lowest precedence
                    SemanticContext.PrecedencePredicate reduced = precedencePredicates.Min();
                    operands.Add(reduced);
                }
                opnds = operands.ToArray();
            }

            public override ICollection<SemanticContext> Operands
            {
                get
                {
                    return Arrays.AsList(opnds);
                }
            }

            public override bool Equals(object obj)
            {
                if (this == obj)
                {
                    return true;
                }
                if (!(obj is SemanticContext.AND))
                {
                    return false;
                }
                SemanticContext.AND other = (SemanticContext.AND)obj;
                return Arrays.Equals(this.opnds, other.opnds);
            }

            public override int GetHashCode()
            {
                return MurmurHash.HashCode(opnds, typeof(SemanticContext.AND).GetHashCode());
            }

            public override bool Eval<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            {
                foreach (SemanticContext opnd in opnds)
                {
                    if (!opnd.Eval(parser, parserCallStack))
                    {
                        return false;
                    }
                }
                return true;
            }

            public override SemanticContext EvalPrecedence<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            {
                bool differs = false;
                IList<SemanticContext> operands = new List<SemanticContext>();
                foreach (SemanticContext context in opnds)
                {
                    SemanticContext evaluated = context.EvalPrecedence(parser, parserCallStack);
                    differs |= (evaluated != context);
                    if (evaluated == null)
                    {
                        // The AND context is false if any element is false
                        return null;
                    }
                    else
                    {
                        if (evaluated != Empty.Instance)
                        {
                            // Reduce the result by skipping true elements
                            operands.Add(evaluated);
                        }
                    }
                }
                if (!differs)
                {
                    return this;
                }
                if (operands.Count == 0)
                {
                    // all elements were true, so the AND context is true
                    return Empty.Instance;
                }
                SemanticContext result = operands[0];
                for (int i = 1; i < operands.Count; i++)
                {
                    result = SemanticContext.AndOp(result, operands[i]);
                }
                return result;
            }

            public override string ToString()
            {
                return Utils.Join("&&", opnds);
            }
        }

        public class OR : SemanticContext.Operator
        {
            [NotNull]
            public readonly SemanticContext[] opnds;

            public OR(SemanticContext a, SemanticContext b)
            {
                HashSet<SemanticContext> operands = new HashSet<SemanticContext>();
                if (a is SemanticContext.OR)
                {
                    operands.UnionWith(((OR)a).opnds);
                }
                else
                {
                    operands.Add(a);
                }
                if (b is SemanticContext.OR)
                {
                    operands.UnionWith(((OR)b).opnds);
                }
                else
                {
                    operands.Add(b);
                }
                IList<SemanticContext.PrecedencePredicate> precedencePredicates = FilterPrecedencePredicates(operands);
                if (precedencePredicates.Count > 0)
                {
                    // interested in the transition with the highest precedence
                    SemanticContext.PrecedencePredicate reduced = precedencePredicates.Max();
                    operands.Add(reduced);
                }
                this.opnds = operands.ToArray();
            }

            public override ICollection<SemanticContext> Operands
            {
                get
                {
                    return Arrays.AsList(opnds);
                }
            }

            public override bool Equals(object obj)
            {
                if (this == obj)
                {
                    return true;
                }
                if (!(obj is SemanticContext.OR))
                {
                    return false;
                }
                SemanticContext.OR other = (SemanticContext.OR)obj;
                return Arrays.Equals(this.opnds, other.opnds);
            }

            public override int GetHashCode()
            {
                return MurmurHash.HashCode(opnds, typeof(SemanticContext.OR).GetHashCode());
            }

            public override bool Eval<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            {
                foreach (SemanticContext opnd in opnds)
                {
                    if (opnd.Eval(parser, parserCallStack))
                    {
                        return true;
                    }
                }
                return false;
            }

            public override SemanticContext EvalPrecedence<Symbol, ATNInterpreter>(Recognizer<Symbol, ATNInterpreter> parser, RuleContext parserCallStack)
            {
                bool differs = false;
                IList<SemanticContext> operands = new List<SemanticContext>();
                foreach (SemanticContext context in opnds)
                {
                    SemanticContext evaluated = context.EvalPrecedence(parser, parserCallStack);
                    differs |= (evaluated != context);
                    if (evaluated == Empty.Instance)
                    {
                        // The OR context is true if any element is true
                        return Empty.Instance;
                    }
                    else
                    {
                        if (evaluated != null)
                        {
                            // Reduce the result by skipping false elements
                            operands.Add(evaluated);
                        }
                    }
                }
                if (!differs)
                {
                    return this;
                }
                if (operands.Count == 0)
                {
                    // all elements were false, so the OR context is false
                    return null;
                }
                SemanticContext result = operands[0];
                for (int i = 1; i < operands.Count; i++)
                {
                    result = SemanticContext.OrOp(result, operands[i]);
                }
                return result;
            }

            public override string ToString()
            {
                return Utils.Join("||", opnds);
            }
        }

        public static SemanticContext AndOp(SemanticContext a, SemanticContext b)
        {
            if (a == null || a == Empty.Instance)
            {
                return b;
            }
            if (b == null || b == Empty.Instance)
            {
                return a;
            }
            SemanticContext.AND result = new SemanticContext.AND(a, b);
            if (result.opnds.Length == 1)
            {
                return result.opnds[0];
            }
            return result;
        }

        public static SemanticContext OrOp(SemanticContext a, SemanticContext b)
        {
            if (a == null)
            {
                return b;
            }
            if (b == null)
            {
                return a;
            }
            if (a == Empty.Instance || b == Empty.Instance)
            {
                return Empty.Instance;
            }
            SemanticContext.OR result = new SemanticContext.OR(a, b);
            if (result.opnds.Length == 1)
            {
                return result.opnds[0];
            }
            return result;
        }

        private static IList<SemanticContext.PrecedencePredicate> FilterPrecedencePredicates(HashSet<SemanticContext> collection)
        {
            if (!collection.OfType<PrecedencePredicate>().Any())
                Collections.EmptyList<PrecedencePredicate>();

            List<PrecedencePredicate> result = collection.OfType<PrecedencePredicate>().ToList();
            collection.ExceptWith(result);
            return result;
        }
    }
}
